gen/field.go

```diff
@@ -95,22 +104,79 @@ func (f fieldGroupGenerator) DefineStruct(g Generator) error {
 		}`,
 		f,
 		TemplateFunc("tag", func(f *compile.FieldSpec) string {
-			// We want to add omitempty if the field is an optional struct or
-			// primitive to redure "null" noise. We won't add omitempty for
-			// optional collections because omitempty doesn't differentiate
-			// between nil and empty collections.
-
-			if (isStructType(f.Type) || isPrimitiveType(f.Type)) && !f.Required {
-				return fmt.Sprintf("`json:\"%s,omitempty\"`", f.Name)
-			}
-
-			return fmt.Sprintf("`json:%q`", f.Name)
-			// TODO(abg): Take go.tag and js.name annotations into account
+			return generateTags(f)
 		}),
 		TemplateFunc("declFieldName", f.declFieldName),
 	)
 }
 
+// generateTags parses the annotation on the thrift field and creates the resulting go tag
+func generateTags(f *compile.FieldSpec) string {
+	// start with default json tag
+	tags := []string{
+		compileTag(jsonTagName, compileJSONTagValue(f, f.Name)),
+	}
+
+	// process go tags and overwrite json tag if specified in thrift annotation
+	if goAnnotation := f.ThriftAnnotations()[goTagName]; goAnnotation != "" {
+		if jt := reflect.StructTag(goAnnotation).Get(jsonTagName); jt != "" {
+			// truncate current tags to remove current json tag and compile the one defined in thrift
+			goTags := annotationsFromRawTag(goAnnotation)
+			goTags[jsonTagName] = compileJSONTagValue(f, jt)
+			tags, f.Annotations[goTagName] = tags[1:], annotationsToRawTag(goTags)
```

Comment:
Don't mutate the FieldSpec. Codegen logic must not change the compiled representation of the Thrift file.